<img src="../src/packt-banner.png" alt="">

### Exercise 3: Build a SVM modle for Face Recognition Problem
##### (25 points) --> your total will divided by 5 to get 5 points for this exercise.
---

We will use a very famous dataset, called Labelled Faces in the Wild, which
consists of 1288 faces of famous people, and it is available at http://viswww.cs.umass.edu/lfw/lfw-funneled.tgz. 

However, note that it can be easily imported via scikit-learn from the datasets class.
Each image consists of 1850 features: we could proceed by simply using each of them in the model.



Fitting a SVM to non-linear data using the Kernel Trick produces non- linear decision boundaries.
In particular, we seek to:
* Build SVM model with radial basis function (RBF) kernel
* Use a grid search cross-validation to explore ran- dom combinations of parameters.

### Step to do:

1. Loading the dataf from sklearn.datasets:

In [2]:
from sklearn.datasets import fetch_lfw_people
faces = fetch_lfw_people(min_faces_per_person=60)

2. Since the data can be accessed from the sklearn.datasets module, you need to explore the dataset. 
    - (refer to the first 6 steps in Lab_1 could help you)

a- Print the field names (that is, the keys to the dictionary) (1 point)

In [4]:
# What fields are in the dictionary?
field_names = [attr for attr in dir(faces)]
print(field_names)

['DESCR', 'data', 'images', 'target', 'target_names']


b- Print the dataset description contained (2 point)

In [20]:
# write your code here
print(faces.DESCR)

.. _labeled_faces_in_the_wild_dataset:

The Labeled Faces in the Wild face recognition dataset
------------------------------------------------------

This dataset is a collection of JPEG pictures of famous people collected
over the internet, all details are available on the official website:

    http://vis-www.cs.umass.edu/lfw/

Each picture is centered on a single face. The typical task is called
Face Verification: given a pair of two pictures, a binary classifier
must predict whether the two images are from the same person.

An alternative task, Face Recognition or Face Identification is:
given the picture of the face of an unknown person, identify the name
of the person by referring to a gallery of previously seen pictures of
identified persons.

Both Face Verification and Face Recognition are tasks that are typically
performed on the output of a model trained to perform Face Detection. The
most popular model for Face Detection is called Viola-Jones and is
implemented in the OpenC

3. Print the data, its shape, and the target names. ( 3 points)

In [23]:
# What does the data look like?
print(faces.data)

[[0.53464055 0.5254902  0.49673203 ... 0.00653595 0.00653595 0.00261438]
 [0.28627452 0.20784314 0.2522876  ... 0.96993464 0.9490196  0.9346406 ]
 [0.31895426 0.39215687 0.275817   ... 0.4261438  0.7908497  0.9555555 ]
 ...
 [0.11633987 0.11111111 0.10196079 ... 0.5686274  0.5803922  0.5542484 ]
 [0.19346406 0.21176471 0.2901961  ... 0.6862745  0.654902   0.5908497 ]
 [0.12287582 0.09803922 0.10980392 ... 0.12941177 0.1633987  0.29150328]]


In [24]:
# what is the shape of the data?
print(faces.data.shape)

(1348, 2914)


In [26]:
# What is the target names?
print(faces.target_names)

['Ariel Sharon' 'Colin Powell' 'Donald Rumsfeld' 'George W Bush'
 'Gerhard Schroeder' 'Hugo Chavez' 'Junichiro Koizumi' 'Tony Blair']


4. Divide the data into features (X) using the faces.data and target (y) using faces.target (2 points)

In [3]:
# Write your code here 
X = faces.data
y = faces.target

5. Splitting the data into training and testing sets. (2 point)

We train the model with 70% of the samples and test with the remaining 30%.

In [4]:
# Write your code here 
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


# print the sizes of our training and test set to verify if the splitting has occurred properly.
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)


(943, 2914)
(405, 2914)
(943,)
(405,)


6. Declare SVM model with kernel='rbf', class_weight='balanced' (2 points)

In [5]:
from sklearn.svm import SVC

# write your code here
model = SVC(kernel='rbf', class_weight='balanced')

7. Use a grid search cross-validationwith 10 CV to explore random combinations of parameters. (3 points) 
    - we will adjust C, which controls the margin
    - and Gamma (γ), which controls the size of the radial basis function kernel, and determine the best model.

In [6]:
from sklearn.model_selection import GridSearchCV
param_grid = {'C': [1,5,10,50],'gamma': [0.001,0.0005,0.01,0.1]}
# write your code here for GridSearchCV:
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=10)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=10, estimator=SVC(class_weight='balanced'),
             param_grid={'C': [1, 5, 10, 50],
                         'gamma': [0.001, 0.0005, 0.01, 0.1]})

8. predict on the test set, using the best model from above step (best_estimator_) (5 points)

In [8]:
# write your code here
best_estimator = grid_search.best_estimator_
print("Best estimator:", best_estimator)
best_estimator.fit(X_train,y_train)
y_pred = best_estimator.predict(X_test)

Best estimator: SVC(C=50, class_weight='balanced', gamma=0.001)


9. Model performances:
Run the following code to print the model evaluation metric

In [9]:
from sklearn.metrics import classification_report
labels = list(faces.target_names)
print(classification_report(y_test,y_pred,target_names=labels))

                   precision    recall  f1-score   support

     Ariel Sharon       0.59      0.76      0.67        17
     Colin Powell       0.80      0.83      0.81        84
  Donald Rumsfeld       0.67      0.81      0.73        36
    George W Bush       0.90      0.86      0.88       146
Gerhard Schroeder       0.69      0.71      0.70        28
      Hugo Chavez       1.00      0.63      0.77        27
Junichiro Koizumi       0.89      1.00      0.94        16
       Tony Blair       0.82      0.78      0.80        51

         accuracy                           0.81       405
        macro avg       0.79      0.80      0.79       405
     weighted avg       0.83      0.81      0.82       405



10. What do you observe about the model performances? (5 points)

#### Write your answer here




The model has an accuracy rate of 0.81. 
The f-1 score (which is a good indicator for correctly predicted positive for a dataset) shows some deviation for every target.
As we can see model obtains a 0.67 score for Ariel Sharon, which is not very impressive, this could be due to small sample which we can see from the support field.
On the other hand model predicted Junichiro's data accurately 0.94 times which is very decent and on top it had even less data to work with, so the reason for the discrepency between the score for these targets could be that the data for Ariel was scattered much more than any other target causing model to miss much more often.